src/openrct2/world/Balloon.cpp

```diff
@@ -0,0 +1,143 @@
+* (at your option) any later version.
+*
+* A full copy of the GNU General Public License can be found in licence.txt
+*****************************************************************************/
+#pragma endregion
+
+#include "../network/network.h"
+
+extern "C"
+{
+    #include "../audio/audio.h"
+    #include "../scenario/scenario.h"
+    #include "../util/util.h"
+    #include "sprite.h"
+}
+
+bool rct_sprite::IsBalloon()
+{
+    return this->balloon.sprite_identifier == SPRITE_IDENTIFIER_MISC &&
+           this->balloon.misc_identifier == SPRITE_MISC_BALLOON;
+}
+
+rct_balloon * rct_sprite::AsBalloon()
+{
+    rct_balloon * result = nullptr;
+    if (IsBalloon())
+    {
+        result = (rct_balloon *)this;
+    }
+    return result;
+}
+
+void rct_balloon::Update()
+{
+    invalidate_sprite_2((rct_sprite *)this);
+    if (popped == 1)
+    {
+        frame += 256;
+        if (frame >= 1280)
+        {
+            sprite_remove((rct_sprite *)this);
+        }
+    }
+    else
+    {
+        sint32 original_var26a = var_26a;
+        var_26a += 85;
+        if (original_var26a >= 255 - 85)
+        {
+            var_26b++;
+            sprite_move(x, y, z + 1, (rct_sprite*)this);
+
+            sint32 maxZ = 1967 - ((x ^ y) & 31);
+            if (z >= maxZ)
+            {
+                Pop();
+            }
+        }
+    }
+}
+
+void rct_balloon::Press()
+{
+    if (popped != 1)
+    {
+        // There is a random chance that pressing the balloon will not pop it
+        // and instead shift it slightly
+        uint32 random = scenario_rand();
+        if ((sprite_index & 7) || (random & 0xFFFF) < 0x2000)
+        {
+            Pop();
+        }
+        else
+        {
+            sint16 shift = ((random & 0x80000000) ? -6 : 6);
+            sprite_move(x + shift, y, z, (rct_sprite *)this);
+        }
+    }
+}
+
+void rct_balloon::Pop()
+{
+    popped = 1;
+    frame = 0;
+    audio_play_sound_at_location(SOUND_BALLOON_POP, x, y, z);
+}
+
+extern "C"
+{
+    void create_balloon(sint32 x, sint32 y, sint32 z, sint32 colour, bool isPopped)
+    {
+        rct_sprite* sprite = create_sprite(2);
+        if (sprite != nullptr)
+        {
+            sprite->balloon.var_14 = 13;
+            sprite->balloon.var_09 = 22;
+            sprite->balloon.var_15 = 11;
+            sprite->balloon.sprite_identifier = SPRITE_IDENTIFIER_MISC;
+            sprite_move(x, y, z, sprite);
+            sprite->balloon.misc_identifier = SPRITE_MISC_BALLOON;
+            sprite->balloon.frame = 0;
+            sprite->balloon.colour = colour;
+            sprite->balloon.popped = (isPopped ? 1 : 0);
+        }
+    }
+
+    void balloon_update(rct_balloon * balloon)
+    {
+        balloon->Update();
+    }
+
+    void game_command_balloon_press(sint32 * eax, sint32 * ebx, sint32 * ecx, sint32 * edx, sint32 * esi, sint32 * edi, sint32 * ebp)
+    {
+        uint16 spriteIndex = *eax & 0xFFFF;
```

Comment:
I find it much nicer to read when the game command is just a stub and it calls a separate function. That way its much easier to know what registers are passed into and out of a command.